gensim/models/poincare.py

```diff
@@ -0,0 +1,637 @@
+        self.np_random = np_random.RandomState(seed)
+        self.init_range = (-0.001, 0.001)
+        self.loss_grad = grad(PoincareModel.loss_fn)
+        self.load_relations()
+        self.init_embeddings()
+
+    def load_relations(self):
+        """Load relations from the train file and build vocab."""
+        vocab = {}
+        index2word = []
+        all_relations = []  # List of all relation pairs
+        term_relations = defaultdict(set)  # Mapping from node index to its related node indices
+
+        with smart_open(self.train_file, 'r', encoding=self.encoding) as f:
+            reader = csv.reader(f, delimiter='\t')
+            for row in reader:
+                assert len(row) == 2, 'Relation pair has more than two items'
+                for item in row:
+                    if item in vocab:
+                        vocab[item].count += 1
+                    else:
+                        vocab[item] = Vocab(count=1, index=len(index2word))
+                        index2word.append(item)
+                node_1, node_2 = row
+                node_1_index, node_2_index = vocab[node_1].index, vocab[node_2].index
+                term_relations[node_1_index].add(node_2_index)
+                relation = (node_1_index, node_2_index)
+                all_relations.append(relation)
+
+        self.wv.vocab = vocab
+        self.wv.index2word = index2word
+        self.indices_set = set((range(len(index2word))))  # Set of all node indices
+        self.indices_array = np.array(range(len(index2word)))  # Numpy array of all node indices
+        counts = np.array([self.wv.vocab[index2word[i]].count for i in range(len(index2word))])
+        self.node_probabilities = counts / counts.sum()
+        self.all_relations = all_relations
+        self.term_relations = term_relations
+        self.negatives_buffer = []  # Buffer to store negative samples, to reduce calls to sampling method
+        self.negatives_buffer_index = 0  # Position in buffer till which samples have been consumed
+        self.negatives_buffer_size = 2000
+
+    def init_embeddings(self):
+        """Randomly initialize vectors for the items in the vocab."""
+        shape = (len(self.wv.index2word), self.size)
+        self.wv.syn0 = self.np_random.uniform(self.init_range[0], self.init_range[1], shape)
+
+    def get_candidate_negatives(self):
+        """
+        Returns candidate negatives of size `self.negative` from the negative examples buffer.
+
+        Returns
+        --------
+        numpy array
+            numpy array of shape (`self.negative`,) containing indices of negative nodes.
+        """
+
+        if self.negatives_buffer_index >= len(self.negatives_buffer):
+            # Note: np.random.choice much slower than random.sample for large populations, possible bottleneck
+            self.negatives_buffer = self.np_random.choice(self.indices_array, size=self.negatives_buffer_size, p=self.node_probabilities)
```

Comment:
line too long

Reply:
Fixed (along with a minor change in logic)